File: backend/src/services/graph_contracts.py
"""graph_contracts — Decision graph integrity validation.

Enforces the run → decision → outcome → value lineage chain.
Called at API boundaries (create, execute, compute) to guarantee that
every entity has valid upward links before it enters the system.

Contract rules:
  Decision:  decision_id exists, source_run_id resolves, status is valid
  Outcome:   outcome_id exists, source_decision_id resolves to existing decision
  Value:     source_outcome_id resolves, linked outcome has source_decision_id,
             linked decision exists — never a free-floating run-only value

All validators return (ok: bool, error: str | None).
Callers decide whether to reject (HTTP 400/422) or warn-and-proceed.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

logger = logging.getLogger(__name__)

# Valid statuses — must match schemas/operator_decision.py Literals exactly
VALID_DECISION_STATUSES = frozenset({
    "CREATED", "IN_REVIEW", "EXECUTED", "FAILED", "CLOSED",
})

VALID_OUTCOME_STATUSES = frozenset({
    "PENDING_OBSERVATION", "OBSERVED", "CONFIRMED", "DISPUTED", "CLOSED", "FAILED",
})


@dataclass(frozen=True)
class ContractResult:
    ok: bool
    error: str | None = None


# ── Decision contracts ──────────────────────────────────────────────────────

def validate_decision_create(body: dict) -> ContractResult:
    """Validate decision creation request has valid graph linkage.

    Rules:
      - source_run_id, if provided, must resolve to an existing run
      - decision_type must be one of the schema-defined types
    """
    from src.services import run_store

    run_id = body.get("source_run_id")
    if run_id:
        run = run_store.get(run_id)
        if run is None:
            return ContractResult(
                ok=False,
                error=f"source_run_id '{run_id}' does not resolve to an existing run",
            )

    return ContractResult(ok=True)


def validate_decision_status(status: str) -> ContractResult:
    """Validate that a decision status is in the allowed set."""
    if status not in VALID_DECISION_STATUSES:
        return ContractResult(
            ok=False,
            error=f"Invalid decision_status '{status}' — expected one of {sorted(VALID_DECISION_STATUSES)}",
        )
    return ContractResult(ok=True)


def validate_decision_execute(decision: dict) -> ContractResult:
    """Pre-execute validation: decision must have outcome_id linkage."""
    if not decision.get("outcome_id"):
        return ContractResult(
            ok=False,
            error=f"Decision {decision.get('decision_id')} has no outcome_id — cannot execute without linked outcome",
        )
    return ContractResult(ok=True)


# ── Outcome contracts ───────────────────────────────────────────────────────

def validate_outcome_create(body: dict) -> ContractResult:
    """Validate outcome creation request has valid graph linkage.

    Rules:
      - source_decision_id, if provided, must resolve to an existing decision
      - the linked decision must have a valid source_run_id
    """
    from src.services import decision_operator_store

    decision_id = body.get("source_decision_id")
    if decision_id:
        decision = decision_operator_store.get(decision_id)
        if decision is None:
            return ContractResult(
                ok=False,
                error=f"source_decision_id '{decision_id}' does not resolve to an existing decision",
            )
        if not decision.get("source_run_id"):
            logger.warning(
                "Outcome created for decision %s which has no source_run_id — lineage incomplete",
                decision_id,
            )

    return ContractResult(ok=True)


def validate_outcome_status(status: str) -> ContractResult:
    """Validate that an outcome status is in the allowed set."""
    if status not in VALID_OUTCOME_STATUSES:
        return ContractResult(
            ok=False,
            error=f"Invalid outcome_status '{status}' — expected one of {sorted(VALID_OUTCOME_STATUSES)}",
        )
    return ContractResult(ok=True)


# ── Value contracts ─────────────────────────────────────────────────────────

def validate_value_chain(outcome: dict) -> ContractResult:
    """Full-chain validation for value computation.

    Rules:
      - outcome must have source_decision_id (not free-floating)
      - the linked decision must exist
      - the linked decision must have source_run_id
    """
    from src.services import decision_operator_store

    decision_id = outcome.get("source_decision_id")
    if not decision_id:
        return ContractResult(
            ok=False,
            error=(
                f"Outcome '{outcome.get('outcome_id')}' has no source_decision_id — "
                "value cannot be computed from an unlinked outcome"
            ),
        )

    decision = decision_operator_store.get(decision_id)
    if decision is None:
        return ContractResult(
            ok=False,
            error=(
                f"Outcome '{outcome.get('outcome_id')}' references decision '{decision_id}' "
                "which does not exist — graph is broken"
            ),
        )

    if not decision.get("source_run_id"):
        logger.warning(
            "Value computed for decision %s with no source_run_id — lineage incomplete but allowed",
            decision_id,
        )

    return ContractResult(ok=True)


# ── Post-mutation verification ──────────────────────────────────────────────

def verify_decision_outcome_linkage(decision: dict) -> ContractResult:
    """After create or execute, verify bidirectional linkage is intact.

    Rules:
      - decision.outcome_id must be non-null
      - outcome with that ID must exist
      - outcome.source_decision_id must point back to this decision
    """
    from src.services import outcome_store

    outcome_id = decision.get("outcome_id")
    if not outcome_id:
        return ContractResult(
            ok=False,
            error=f"Decision {decision.get('decision_id')} has no outcome_id after linkage",
        )

    outcome = outcome_store.get(outcome_id)
    if outcome is None:
        return ContractResult(
            ok=False,
            error=f"Decision {decision.get('decision_id')} references outcome {outcome_id} which does not exist",
        )

    if outcome.get("source_decision_id") != decision.get("decision_id"):
        return ContractResult(
            ok=False,
            error=(
                f"Bidirectional linkage broken: decision {decision.get('decision_id')} → "
                f"outcome {outcome_id}, but outcome.source_decision_id = "
                f"'{outcome.get('source_decision_id')}'"
            ),
        )

    return ContractResult(ok=True)

File: frontend/src/lib/graph-contracts.ts
/**
 * graph-contracts — Frontend read-path contract enforcement.
 *
 * Mirrors backend src/services/graph_contracts.py logic.
 * Applied at Zustand store sync points so all persona views
 * receive only contract-valid records.
 *
 * Fail closed: orphan records are silently dropped with a
 * single console.warn per orphan (minimal noise, auditable).
 */

import type {
  OperatorDecision,
  OperatorDecisionStatus,
  Outcome,
  OutcomeLifecycleStatus,
  DecisionValue,
} from "@/types/observatory";

// ── Valid status sets (mirrors backend VALID_* constants) ─────────────────────

const VALID_DECISION_STATUSES = new Set<OperatorDecisionStatus>([
  "CREATED",
  "IN_REVIEW",
  "EXECUTED",
  "FAILED",
  "CLOSED",
]);

const VALID_OUTCOME_STATUSES = new Set<OutcomeLifecycleStatus>([
  "PENDING_OBSERVATION",
  "OBSERVED",
  "CONFIRMED",
  "DISPUTED",
  "CLOSED",
  "FAILED",
]);

// ── Record validators ─────────────────────────────────────────────────────────

/**
 * A decision is valid if it has a non-empty id and a recognised status.
 * source_run_id absence is allowed (warn only — creation path may omit it).
 */
export function validateDecisionRecord(d: OperatorDecision): boolean {
  if (!d.decision_id) return false;
  if (!VALID_DECISION_STATUSES.has(d.decision_status)) return false;
  return true;
}

/**
 * An outcome is valid if it has a non-empty id, a recognised status,
 * and its source_decision_id resolves to an entry in the provided decision set.
 */
export function validateOutcomeRecord(
  o: Outcome,
  decisionIds: Set<string>,
): boolean {
  if (!o.outcome_id) return false;
  if (!VALID_OUTCOME_STATUSES.has(o.outcome_status)) return false;
  // Outcomes must be linked to a known decision.
  if (!o.source_decision_id || !decisionIds.has(o.source_decision_id))
    return false;
  return true;
}

/**
 * A value is valid if it has a non-empty id, its source_outcome_id resolves
 * to a known outcome, and the chain continues to a known decision.
 */
export function validateValueRecord(
  v: DecisionValue,
  outcomeIds: Set<string>,
  decisionIds: Set<string>,
): boolean {
  if (!v.value_id) return false;
  if (!v.source_outcome_id || !outcomeIds.has(v.source_outcome_id))
    return false;
  // Full chain: value → outcome → decision
  if (!v.source_decision_id || !decisionIds.has(v.source_decision_id))
    return false;
  return true;
}

// ── Filter helpers ────────────────────────────────────────────────────────────

/**
 * Return only contract-valid decisions. Orphans are warned and dropped.
 */
export function filterValidDecisions(
  decisions: OperatorDecision[],
): OperatorDecision[] {
  const valid: OperatorDecision[] = [];
  for (const d of decisions) {
    if (validateDecisionRecord(d)) {
      valid.push(d);
    } else {
      console.warn(
        "[graph-contracts] orphan decision dropped",
        d.decision_id,
        d.decision_status,
      );
    }
  }
  return valid;
}

/**
 * Return only contract-valid outcomes. Requires the filtered decision list
 * (already validated) to check parent linkage.
 */
export function filterValidOutcomes(
  outcomes: Outcome[],
  decisions: OperatorDecision[],
): Outcome[] {
  const decisionIds = new Set(decisions.map((d) => d.decision_id));
  const valid: Outcome[] = [];
  for (const o of outcomes) {
    if (validateOutcomeRecord(o, decisionIds)) {
      valid.push(o);
    } else {
      console.warn(
        "[graph-contracts] orphan outcome dropped",
        o.outcome_id,
        o.source_decision_id,
      );
    }
  }
  return valid;
}

/**
 * Return only contract-valid decision values. Requires filtered decisions
 * and outcomes to verify the full chain.
 */
export function filterValidValues(
  values: DecisionValue[],
  outcomes: Outcome[],
  decisions: OperatorDecision[],
): DecisionValue[] {
  const outcomeIds  = new Set(outcomes.map((o) => o.outcome_id));
  const decisionIds = new Set(decisions.map((d) => d.decision_id));
  const valid: DecisionValue[] = [];
  for (const v of values) {
    if (validateValueRecord(v, outcomeIds, decisionIds)) {
      valid.push(v);
    } else {
      console.warn(
        "[graph-contracts] orphan value dropped",
        v.value_id,
        v.source_outcome_id,
        v.source_decision_id,
      );
    }
  }
  return valid;
}
